### PR TITLE
Hue translators with promises, TypeScript

### PR DESCRIPTION
**Pull request by jasongin**
Adding 2 variations of the Hue lamp translator. The "js-promise" variation
returns promises from the methods for better async benavior. The "ts"
variation is also async and uses TypeScript for type-checking and async/await
with promises. (Async translator methods require bridge changes that are
not yet merged.)

### CODE REVIEW

File: org.OpenT2T.Sample.SuperPopular.Lamp/Philips Hue Light Bulb/ts/test.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+var translator = require('./thingTranslator.js');
+var q = require('q');
```
Comment (tjaffri) [inline org.OpenT2T.Sample.SuperPopular.Lamp/Philips Hue Light Bulb/ts/test.js]: I was hoping the test file could also be TypeScript, so we could show direct consumers of this translator how to call it from typescript (e.g. if somebody is writing a bot or website in ts they could refer to the test.ts file to see how to call it). Any chance you could update this file to ts?

File: org.OpenT2T.Sample.SuperPopular.Lamp/Philips Hue Light Bulb/ts/thingTranslator.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
+
+declare function require(module: string): any;
```
Comment (tjaffri) [inline org.OpenT2T.Sample.SuperPopular.Lamp/Philips Hue Light Bulb/ts/thingTranslator.ts]: shouldn't this be an import, not a require? I am not a typescript expert, but I thought that's how you do it in typescript? See https://basarat.gitbooks.io/typescript/content/docs/quick/nodejs.html

Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Lamp/Philips Hue Light Bulb/ts/thingTranslator.ts]: Import doesn't work when there are no type definitions. I could have created (partial) type definitions for node-hue-api, but I wanted to demonstrate how that isn't necessary.

**Comment by jasongin** [comment]
I added a commit which converts test.js to TypeScript, and automatically sets up typings for a few things: require, q, optimist.

**Pull request merged**